persist/fs/seek.go

```diff
@@ -0,0 +1,247 @@
+		filesetPathFromTime(shardDir, blockStart, indexFileSuffix):  &indexFd,
+		filesetPathFromTime(shardDir, blockStart, dataFileSuffix):   &dataFd,
+		filesetPathFromTime(shardDir, blockStart, digestFileSuffix): &digestFd,
+	}); err != nil {
+		return err
+	}
+	s.infoFdWithDigest.Reset(infoFd)
+	s.indexFdWithDigest.Reset(indexFd)
+	s.digestFdWithDigestContents.Reset(digestFd)
+	if err := s.readDigest(); err != nil {
+		// Try to close if failed to read
+		s.Close()
+		return err
+	}
+	infoStat, err := infoFd.Stat()
+	if err != nil {
+		s.Close()
+		return err
+	}
+	indexStat, err := indexFd.Stat()
+	if err != nil {
+		s.Close()
+		return err
+	}
+	if err := s.readInfo(int(infoStat.Size())); err != nil {
+		s.Close()
+		return err
+	}
+	if err := s.readIndex(int(indexStat.Size())); err != nil {
+		s.Close()
+		return err
+	}
+
+	s.dataFd = dataFd
+	return nil
+}
+
+func (s *seeker) prepareUnreadBuf(size int) {
+	if len(s.unreadBuf) < size {
+		s.unreadBuf = make([]byte, size)
+	}
+}
+
+func (s *seeker) readDigest() error {
+	var err error
+	if s.expectedInfoDigest, err = s.digestFdWithDigestContents.ReadDigest(); err != nil {
+		return err
+	}
+	if s.expectedIndexDigest, err = s.digestFdWithDigestContents.ReadDigest(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *seeker) readInfo(size int) error {
+	s.prepareUnreadBuf(size)
+	n, err := s.infoFdWithDigest.ReadAllAndValidate(s.unreadBuf[:size], s.expectedInfoDigest)
+	if err != nil {
+		return err
+	}
+	info, err := readInfo(s.unreadBuf[:n])
+	if err != nil {
+		return err
+	}
+	s.start = xtime.FromNanoseconds(info.Start)
+	s.blockSize = time.Duration(info.BlockSize)
+	s.entries = int(info.Entries)
+	return nil
+}
+
+func (s *seeker) readIndex(size int) error {
+	s.prepareUnreadBuf(size)
+	indexBytes := s.unreadBuf[:size]
+	n, err := s.indexFdWithDigest.ReadAllAndValidate(indexBytes, s.expectedIndexDigest)
+	if err != nil {
+		return err
+	}
+
+	indexUnread := s.unreadBuf[:n][:]
+	s.indexMap = make(map[string]*schema.IndexEntry, s.entries)
+	// Read all entries of index
+	for read := 0; read < s.entries; read++ {
+		entry := &schema.IndexEntry{}
+		size, consumed := proto.DecodeVarint(indexUnread)
+		indexUnread = indexUnread[consumed:]
+		if consumed < 1 {
+			return errReadIndexEntryZeroSize
+		}
+
+		indexEntryData := indexUnread[:size]
+		if err := proto.Unmarshal(indexEntryData, entry); err != nil {
+			return err
+		}
+
+		indexUnread = indexUnread[size:]
+		s.indexMap[string(entry.Id)] = entry
+	}
+
+	return nil
+}
+
+func (s *seeker) Seek(id string) ([]byte, error) {
+	entry, exists := s.indexMap[id]
+	if !exists {
+		return nil, errSeekIDNotFound
+	}
+
+	_, err := s.dataFd.Seek(entry.Offset, 0)
+	if err != nil {
+		return nil, err
+	}
+	s.dataReader.Reset(s.dataFd)
+
+	expectedSize := markerLen + idxLen + int(entry.Size)
+	var data []byte
+	if s.bytesPool != nil {
+		data = s.bytesPool.Get(expectedSize)[:expectedSize]
+	} else {
+		data = make([]byte, expectedSize)
+	}
```

Comment:
Need to do similar thing to the file set reader and use a separate prologue buffer.  Otherwise the pointer we return from this method won't be from the bytes pool.  e.g. returning `data[markerLen+idxLen:]` instead of returning `data`